### DISCUSSION TITLE
Allow machines to share a thread

### DISCUSSION THREAD

**Discussion by winksaville** [AletheiaWareLLC/Wink#48] [Ideas]
(Fixed typo in title)

I started to create an issue, but it might turn out there are possibly several issues, so created a discussion thread instead.

Currently MaCHines (MCH)s are each running in their own thread, long term multiple MCHs should be able to share a single thread. The primary reasons for doing this is performance, efficiency and reduction in resources required for an MCH.

I believe #39 needs to be completed first, but this feature will add a requirement to #39 that "addressing" needs to be independent of a thread when multiple MCHs are sharing a single thread. This could be accomplished by requiring each message to contain a MchId. This MchId could be converted by the runtime to a Machine::handleMessage method to process a message. In addition this MchId needs to be "universal" so that a particular MCH could be sent messages regardless of its location in the [interplanetary internet](https://en.wikipedia.org/wiki/Interplanetary_Internet) (IPIN) which might use InterPaNet (IPN) protocol :)

Joking aside, the MchId might also have an associated "typical" Round Trip Time (RTT) so timeouts can be set appropriately and when an another MchId is looking for an MCH that it considers close enough defined by an RTT range.

**Comment by stuartmscott** [comment]
Feels like a violation of Principle 1: Isolated Fault Units.

Though a future work item is to get WinkServer running on embedded devices, so there'll have to be a single thread of execution that runs through the server and then sequentially through every living machine spawned on that host.

Need to think about;
- how worker tasks, like the one in `time/Ticker`, get scheduled.
- how filesystem operations, like the serving and log directories, are implemented.
- how errors and exit signals are captured to stop a single machine bringing down the whole server including all other machines.

> Interplanetary Internet

Haha something we'll have to think about soon. Round Trip Time Estimate (RTTE) is a nice idea, ideally handled by the networking layer.

**Reply by winksaville** [reply]
> Feels like a violation of Principle 1: Isolated Fault Units.

How so?

I'm proposing the designer decides. Also, if a particular Mch is miss behaving the monitors should still be able to stop, restart, or reload a same or different version. I don't think the granularity is threads or processes, although that's good starting point.

Note, I envision a "system" containing 1,000s or even millions of Mchs so that scale is impractical if the granularity is thread or process. Also the message throughput and RTT is significantly slower between threads as compared to single thread.

To be clear if there are 1,000s or "millions" of Mchs then most are idle most of the time :)
